### PR TITLE
https: soWatcher, shared_libraries use a pathIdentifier as key of ELF binaries

### PR DESCRIPTION
**Pull request by nplanel**
### What does this PR do?

* pathIdentifier is an unique id of and ELF (system-wide) as it contain dev and inode as key.
* New Unregister path, thanks to processMonitor that recieve process EXIT event and unregister the uprobe (maintained by a refcount)
* ebpf UID use pathIdentifier as source of truth and use wider alphabet (base64), specially because the UID is limited (5 chars)

### Motivation

Follow up on [#incident-16860](https://dd.slack.com/archives/C044LV5NP1T), [#incident-18347 ](https://dd.slack.com/archives/C04JV29KQUS)

### Describe how to test/QA your changes

```
system_probe_config:
  log_level: debug
  enable_runtime_compiler: true
network_config:
  enable_http_monitoring: true
  enable_https_monitoring: true
```

1/ run system-probe on Ubuntu 22.04 
 * In a container, run wget and or curl but keep them stopped (Ctrl-Z) (to prefetch the libssl.so/crypto.so/gnutls.so loaded)
 * `wget https://httpbin.org/anything/foo/200`
 * `sudo curl -s --unix-socket /opt/datadog-agent/run/sysprobe.sock http://unix/network_tracer/connections | tee cnx | jq '.conns[].tags , .tags'`
You should see openssl tags : `tls.library:openssl`
 * `sudo curl -s --unix-socket /opt/datadog-agent/run/sysprobe.sock http://unix/network_tracer/debug/http_monitoring`

2/ on staging or other cluster
 * create an image (you can reuse the same image)
 * enable https_monitoring like in [PR](https://github.com/DataDog/k8s-datadog-agent-ops/pull/1572/files)
 * deploy in in staging (oddish-a,b,c)
 * look at USM correctness dashboard
   *  you shall see more https traffic `ingress-haproxy-web`
 * restart some ingress-haproxy pods (ask on #fabric)
    * look if any pods are stuck in Restarting state
    * could be helpful
       *  `watch -n 10 'kubectl --context=oddish-a.us1.staging.dog get pod -o wide -A | grep Terminating'`
       * [incident dashboard](https://ddstaging.datadoghq.com/notebook/4121966/-incident-17876?cell_id=lhiycj1d&range=172800000&tpl_var_datacenter=us1.staging.dog&tpl_var_kube_cluster_name=oddish-%2A&from_ts=1670343653539&start=1672930020000&to_ts=1670516453539&live=false)

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Comment by nplanel** [comment]
moving to 7.43 as the fix will be more complex

**Comment by guyarb** [comment]
After the 2 incident with had with that solution, I think we should describe a QA steps & remove `skip-qa` tag

File: pkg/network/protocols/http/shared_libraries.go
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:102]:
```suggestion
	blockListByID map[pathIdentifier]struct{}
```

File: pkg/network/protocols/http/shared_libraries.go
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:128]: is it still relevant?

File: pkg/network/protocols/http/shared_libraries.go
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:195]: calling `WithAllProcs` (sync) and only then `w.processMonitor.Initialize()` means that we can loose processes spawned during that time

**Review by guyarb — CHANGES_REQUESTED** [review]

File: pkg/network/protocols/http/ebpf_ssl.go
Comment (nplanel) [reply pkg/network/protocols/http/ebpf_ssl.go:363]: it will be associated with path, like `BinaryPath: root+path` was writtening a

File: pkg/network/protocols/http/shared_libraries.go
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:56]: b is just local scratch pad here

File: pkg/network/protocols/http/shared_libraries.go
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:128]: yes we don't support upgrade
I created a jira ticket https://datadoghq.atlassian.net/browse/USMO-172?atlOrigin=eyJpIjoiZGUzMGUyY2MzMWRkNDFjZGE3YjBhZWJkM2YyMDU3YzciLCJwIjoiaiJ9

File: pkg/network/tracer/tracer_usm_linux_test.go
Comment (nplanel) [inline pkg/network/tracer/tracer_usm_linux_test.go:162]: This issue here is that we scan to slowy the ELF, same issue than goTLS

File: pkg/network/protocols/http/shared_libraries.go
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:162]: why do you ignore the error?

File: pkg/network/protocols/http/shared_libraries.go
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:189]: if we matched a rule, is there a reason to keep checking the other rules?

File: pkg/network/protocols/http/shared_libraries_test.go
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries_test.go:77]: Use indicative names so other people can easily read the code
```suggestion
	tempDir := t.TempDir()
```

File: pkg/network/protocols/http/shared_libraries_test.go
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries_test.go:114]: it the code is not relevant, remove it
if it is relevant, uncomment

File: pkg/network/protocols/http/shared_libraries_test.go
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries_test.go:116]: is it useful for the tests?

File: pkg/network/protocols/http/shared_libraries_test.go
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries_test.go:117]: you are mixing between `require.NoError` and `assert.NoError`
please stick with one of them

File: pkg/network/protocols/http/shared_libraries_test.go
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries_test.go:87]: that's not really "sleep", the name is super confusing and does not give the ability to understand the flow of the test

File: pkg/network/protocols/http/ebpf_ssl.go
Comment (guyarb) [inline pkg/network/protocols/http/ebpf_ssl.go:388]:
```suggestion
					BinaryPath:              filepath.Join(root, path),
```

File: pkg/network/protocols/http/shared_libraries.go
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:57]: please use meaningful names for the variables

File: pkg/network/protocols/http/ebpf_ssl.go
Comment (guyarb) [inline pkg/network/protocols/http/ebpf_ssl.go:432]: why 5?

File: pkg/network/protocols/http/shared_libraries.go
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:213]: what is that todo? is it still relevant? do you plan to add it as part of the current PR?

File: pkg/network/protocols/http/shared_libraries.go
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:205]: why do you register only exit callback?

File: pkg/network/protocols/http/shared_libraries.go
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:230]:
```suggestion
					path    := lib.Bytes()
					libPath := string(path)
					procPid := fmt.Sprintf("%s/%d", w.procRoot, lib.Pid)
					root    := procPid + "/root"
					cwd     := procPid + "/cwd"
```

File: pkg/network/protocols/http/shared_libraries.go
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:232]: why do you skip?
the logic is unclear

File: pkg/network/protocols/http/shared_libraries.go
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:240]:
```suggestion
					// use cwd of the process as root if the path is relative
```

File: pkg/network/protocols/http/shared_libraries.go
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:284]: should be after locking the mutex

File: pkg/network/protocols/http/shared_libraries.go
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:305]: that's is an always increasing memory

File: pkg/network/protocols/http/shared_libraries.go
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:302]: why do we need to unregister a callback that failed to register?

**Review by guyarb — CHANGES_REQUESTED** [review]

File: pkg/network/protocols/http/shared_libraries_test.go
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries_test.go:87]: It is sleep, busybox change his personality based on the `basename(args[0])` this really useful in initrd or embedded environment (see busybox --install creating symlinks)
I remove the /sleep binary as it's not used directly, but sleep will be called from builtin ash command

File: pkg/network/protocols/http/shared_libraries.go
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:61]: I'm using this form as I'm not creating a pointer of pathIdentifier as pathIdentifier object is comparable and is used mostly as a map key

File: pkg/network/protocols/http/shared_libraries.go
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:284]: thanks

File: pkg/network/protocols/http/shared_libraries.go
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:305]: yes but only on problematic path, so this scope is very limited
The only way to cleanup this entry would be the mTime check of the pathID, see jira USMO-172

File: pkg/network/protocols/http/shared_libraries.go
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:302]: the old implementation use this mechanism of cleanup callback
As the register callback can failed to register any uprobe function, and they are registered one at a time in the kernel
Calling the unregister on error would cleanup already registered functions

Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:302]: I added a comment

File: pkg/network/protocols/http/ebpf_ssl.go
Comment (nplanel) [reply pkg/network/protocols/http/ebpf_ssl.go:388]: we control root and path, and the kernel will solve the path himself when the file is used
filepath.Join() is nice but why spending cpu cleanup-ing a special path ?

File: pkg/network/protocols/http/shared_libraries.go
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:57]: done

File: pkg/network/protocols/http/shared_libraries.go
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:133]: done added comment

File: pkg/network/protocols/http/shared_libraries.go
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:138]: the unit test is designed like that (soRule without unregister callback)

File: pkg/network/protocols/http/shared_libraries.go
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:140]: done

File: pkg/network/protocols/http/shared_libraries.go
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:160]: there are no much to do there, I added log.Warn

File: pkg/network/protocols/http/shared_libraries.go
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:187]: in case of multipe library version, but it's a corner case, I moved added a break

File: pkg/network/protocols/http/shared_libraries.go
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:203]: we just need the EXIT process event to unregister the uprobes
uprobes are registered if the lib.so is openat() from the process
I did some test and scanning /proc/pid/smaps at Exec event, the event happen too early (the loader didn't got time to solve and load the .so)

File: pkg/network/protocols/http/shared_libraries.go
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:211]: add cleanup all uprobes, calling all unregisterCB

File: pkg/network/protocols/http/shared_libraries.go
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:228]: done

File: pkg/network/protocols/http/shared_libraries_test.go
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries_test.go:89]: update: sleep need to be present in the path

**Review by xornivore — COMMENTED** [review]
:wave:  @nplanel I've removed the Processes team from reviewers since there doesn't seem to be shared code in this PR, but let me know if there's something you'd like me to take a closer look at 🙇

File: pkg/network/protocols/http/ebpf_ssl.go
Comment (p-lambert) [inline pkg/network/protocols/http/ebpf_ssl.go:393]: can we extract this to a helper function, add unit tests and document it? there is a lot of going on here and the code is a bit hard to follow :sweat:

File: pkg/network/protocols/http/ebpf_ssl.go
Comment (p-lambert) [inline pkg/network/protocols/http/ebpf_ssl.go:305]: probably worth extracting this to a separate PR if this is the fix to the spinlock contention you mentioned

Comment (nplanel) [reply pkg/network/protocols/http/ebpf_ssl.go:305]: https://github.com/DataDog/datadog-agent/pull/15166

File: pkg/network/protocols/http/ebpf_ssl.go
Comment (nplanel) [reply pkg/network/protocols/http/ebpf_ssl.go:393]: https://datadoghq.atlassian.net/browse/USMO-202

**Comment by nplanel** [comment]
@xornivore I think you left a LGTM previously, this initial PR was split in multiple PRs ... github ask you to re-approve even if he can't find the previous comment anymore

File: pkg/network/protocols/http/shared_libraries.go
Comment (guyarb) [reply pkg/network/protocols/http/shared_libraries.go:205]: > I did some test and scanning /proc/pid/smaps at Exec event, the event happen too early (the loader didn't got time to solve and load the .so)

did we fix the issue?

File: pkg/network/protocols/http/shared_libraries.go
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:283]:
```suggestion
		// we need to cleanup our entries as there are no more processes using this ELF
```

**Review by guyarb — CHANGES_REQUESTED** [review]

File: pkg/network/protocols/http/ebpf_ssl.go
Comment (nplanel) [reply pkg/network/protocols/http/ebpf_ssl.go:474]: it's a kind of space left here, as the full string of the eventName format 
`fmt.Sprintf("%s_%.*s_%s_%s", probeType, maxFuncNameLen, funcName, UID, attachPIDstr)`

Comment (nplanel) [reply pkg/network/protocols/http/ebpf_ssl.go:474]: and functionName is variable but with a minimum guarantee of 10 chars

File: pkg/network/protocols/http/shared_libraries.go
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:50]: done

File: pkg/network/protocols/http/shared_libraries.go
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:205]: We are hooking only on the openat("libssl.so") event so we don't have the problem of early Exec

Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:205]: specifically for the Exec event, it's at the user callback to wait if needed, but the callback will be called (guarantee) with a /proc/pid/ existing if a metadata is required (read metadata.Name)

File: pkg/network/protocols/http/shared_libraries.go
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:283]: done

File: pkg/network/tracer/tracer_usm_linux_test.go
Comment (nplanel) [reply pkg/network/tracer/tracer_usm_linux_test.go]: I don't relay on this trick anymore
I simulate an open() of the libs to prehook

**Comment by xornivore** [comment]
> @xornivore I think you left a LGTM previously, this initial PR was split in multiple PRs ... github ask you to re-approve even if he can't find the previous comment anymore
I can remove myself since it appears you no longer need Processes' approval?

File: pkg/network/protocols/http/ebpf_ssl.go
Comment (guyarb) [reply pkg/network/protocols/http/ebpf_ssl.go:474]: so that's what missing from the documentation, please add it

File: pkg/network/protocols/http/shared_libraries.go
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:50]:
```suggestion
// Key is a unique (system wide) TLDR Base64(murmur3.Sum64(device, inode))
```

File: pkg/network/protocols/http/shared_libraries.go
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:52]:
```suggestion
// It composes the device (minor, major) and inode of a file
// murmur is a non-crypto hashing
```

Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:52]: done

File: pkg/network/protocols/http/shared_libraries.go
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:50]: done

File: pkg/network/tracer/tracer_usm_linux_test.go
Comment (nplanel) [reply pkg/network/tracer/tracer_usm_linux_test.go:202]: done

**Review by guyarb — APPROVED** [review]

**Pull request merged**
